subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesDynamicVersionIntegrationTest.groovy

```diff
@@ -0,0 +1,221 @@
+            failure.assertThatCause(containsNormalizedString("Required quality '$requested' and found incompatible value 'qa'"))
+        }
+
+        where:
+        requested | outcome
+        'qa'      | 'succeeds'
+        'canary'  | 'fails'
+    }
+
+    @Unroll("selects the first version which matches the component-level attributes (requested=#requested)")
+    def "selects the first version which matches the component-level attributes"() {
+        given:
+        repository {
+            'org.test:module:1.3' {
+                attribute('quality', 'rc')
+            }
+            'org.test:module:1.2' {
+                attribute('quality', 'rc')
+            }
+            'org.test:module:1.1' {
+                attribute('quality', 'qa')
+            }
+            'org.test:module:1.0' {
+                attribute('quality', 'beta')
+            }
+        }
+        buildFile << """
+            def quality = Attribute.of("quality", String)
+            configurations {
+                conf.attributes.attribute(quality, 'qa')
+            }
+            
+            dependencies {
+                attributesSchema {
+                    attribute(quality)
+                }
+                conf 'org.test:module:$requested'
+            }
+        """
+
+        when:
+        repositoryInteractions {
+            'org.test:module' {
+                expectVersionListing()
+            }
+            'org.test:module:1.3' {
+                expectGetMetadata()
+            }
+            'org.test:module:1.2' {
+                expectGetMetadata()
+            }
+            'org.test:module:1.1' {
+                expectResolve()
+            }
+        }
+
+        then:
+
+        run ':checkDeps'
+        resolve.expectGraph {
+            root(":", ":test:") {
+                edge("org.test:module:${requested}", 'org.test:module:1.1')
+            }
+        }
+
+        where:
+        requested << ["[1.0,)", latestNotation(), "1.+", "1+", "+"]
+    }
+
+    @Unroll("selects the first version which matches the component-level attributes (requested=#requested) using dependency attributes")
+    def "selects the first version which matches the component-level attributes using dependency attributes"() {
+        given:
+        repository {
+            'org.test:module:1.3' {
+                attribute('quality', 'rc')
+            }
+            'org.test:module:1.2' {
+                attribute('quality', 'rc')
+            }
+            'org.test:module:1.1' {
+                attribute('quality', 'qa')
+            }
+            'org.test:module:1.0' {
+                attribute('quality', 'beta')
+            }
+        }
+        buildFile << """
+            def quality = Attribute.of("quality", String)
+
+            configurations {
+                // This test also makes sure that configuration-level attributes are overwritten
```

Comment:
Looking at the comment, I tracked that override logic - unless mistaken - into `DefaultImmutableAttributesFactory.concat` but did not see a unit test explicitly showing that behavior.
And I also noticed that it is missing tests for the recently added `safeConcat`.
These two things might be worth covering in unit tests.

Do not think it should block merging this though, but a follow up is needed.